core/validation/specific_validations/towns_fund_round_four.py

```diff
@@ -500,6 +500,70 @@ def validate_postcodes(workbook: dict[str, pd.DataFrame]) -> list["TownsFundRoun
         ]
 
 
+def validate_current_project_delivery_status_and_upcoming_comms_milestones(
+    workbook: dict[str, pd.DataFrame]
+) -> list["TownsFundRoundFourValidationFailure"] | None:
+    """
+    Validates the Project Progress table in the Round 4 submission workbook.
+
+    Checks for rows with a Project Delivery Status other than "4. Completed" where both
+    the Current Project Delivery Stage or Most Important Upcoming Comms Milestone are null.
+    If such invalid states are found, returns a list of TownsFundRoundFourValidationFailure instances.
+
+    :param workbook: A dictionary containing sheet names as keys and pandas DataFrames
+                     representing each sheet in the Round 4 submission.
+    :return: A list of TownsFundRoundFourValidationFailure instances if validation failures occur,
+             otherwise returns None.
+    """
+    project_progress_df = workbook["Project Progress"]
+
+    project_incomplete_rows = project_progress_df[project_progress_df["Project Delivery Status"] != "4. Completed"]
+
+    columns_to_check = [
+        "Current Project Delivery Stage",
+        "Most Important Upcoming Comms Milestone",
+        "Date of Most Important Upcoming Comms Milestone (e.g. Dec-22)",
+    ]
+
+    invalid_rows = pd.DataFrame()
+
+    for column in columns_to_check:
+        # get the row where an error has occurred with the name of the column saved as 'null_column' for reference
+        invalid_row = find_invalid_rows(project_incomplete_rows, column)
+        # add the row to our final list of invalid rows
+        invalid_rows = pd.concat([invalid_rows, invalid_row], axis=0)
+
+    if len(invalid_rows) > 0:
+        return [
+            TownsFundRoundFourValidationFailure(
+                sheet="Project Progress",
+                section="Projects Progress Summary",
+                column=row["null_column"],
+                message=msgs.BLANK_IF_PROJECT_INCOMPLETE,
+                row_indexes=[row.name],
+            )
+            for _, row in invalid_rows.iterrows()
+        ]
+
```

Comment:
I think creating a new df and doing this concatenation is hard to follow and doesn't offer any benefit.


I propose this way instead (with `find_null_values` in utils):
```    
def validate_current_project_delivery_status_and_upcoming_comms_milestones(
    workbook: dict[str, pd.DataFrame]
) -> list["TownsFundRoundFourValidationFailure"]:
    """
    Validates the Project Progress table in the Round 4 submission workbook.

    Checks for rows with a Project Delivery Status other than "4. Completed" where both
    the Current Project Delivery Stage or Most Important Upcoming Comms Milestone are null.
    If such invalid states are found, returns a list of TownsFundRoundFourValidationFailure instances.

    :param workbook: A dictionary containing sheet names as keys and pandas DataFrames
                     representing each sheet in the Round 4 submission.
    :return: A list of TownsFundRoundFourValidationFailure instances if validation failures occur,
             otherwise returns None.
    """
    project_progress_df = workbook["Project Progress"]

    project_incomplete_rows = project_progress_df[project_progress_df["Project Delivery Status"] != "4. Completed"]

    columns_to_check = [
        "Current Project Delivery Stage",
        "Most Important Upcoming Comms Milestone",
        "Date of Most Important Upcoming Comms Milestone (e.g. Dec-22)",
    ]

    failures = []
    for column in columns_to_check:
        invalid_rows = find_null_values(project_incomplete_rows, column)
        for _, row in invalid_rows.iterrows():
            failures.append(
                TownsFundRoundFourValidationFailure(
                    sheet="Project Progress",
                    section="Projects Progress Summary",
                    column=column,
                    message=msgs.BLANK_IF_PROJECT_INCOMPLETE,
                    row_indexes=[row.name],
                )
            )

    return failures


def find_null_values(df: pd.DataFrame, column: str) -> pd.DataFrame:
    """Helper function to find rows with null values in a specified column.

    :param df: DataFrame to search for invalid rows.
    :param column: Column name to check for null values.
    :return: DataFrame of offending rows
    """
    return df[df[column].isin(NA_VALUES)]
```

Reply:
Makes sense, especially if can refactor leading_factor_of_delay into the cols to check

Reply:
The two functions require slightly different rows to check in and slightly different error messages, but can work if as below:

```
# the column to check alongside the rows it should be checked in and the failure message
    columns_to_check = [
        ("Leading Factor of Delay", project_delayed_rows, msgs.BLANK),
        ("Current Project Delivery Stage", project_incomplete_rows, msgs.BLANK_IF_PROJECT_INCOMPLETE),
        ("Most Important Upcoming Comms Milestone", project_incomplete_rows, msgs.BLANK_IF_PROJECT_INCOMPLETE),
        ("Date of Most Important Upcoming Comms Milestone (e.g. Dec-22)", project_incomplete_rows,
         msgs.BLANK_IF_PROJECT_INCOMPLETE),
    ]

    failures = []
    for column, invalid_rows, message in columns_to_check:
        invalid_rows = find_null_values(invalid_rows, column)
        for _, row in invalid_rows.iterrows():
            failures.append(
                TownsFundRoundFourValidationFailure(
                    sheet="Project Progress",
                    section="Projects Progress Summary",
                    column=column,
                    message=message,
                    row_indexes=[row.name],
                )
            )

    return failures
    ```

Reply:
Looks good